File: app/src/main/java/com/zzzmode/appopsx/ui/core/LangHelper.java
package com.zzzmode.appopsx.ui.core;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import com.zzzmode.appopsx.R;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Created by zl on 2017/6/16.
 */

public class LangHelper {

  private static final String TAG = "LangHelper";

  private static Map<String,Locale> sLocalMap = new HashMap<>();
  private static Locale sDefaultLocal = null;
  static {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      sDefaultLocal = LocaleList.getDefault().get(0);
    }else {
      sDefaultLocal = Locale.getDefault();
    }

    sLocalMap.put("zh-cn",Locale.SIMPLIFIED_CHINESE);
    sLocalMap.put("zh-tw",Locale.TRADITIONAL_CHINESE);
    sLocalMap.put("en",Locale.ENGLISH);
    sLocalMap.put("cs",new Locale("cs","CZ"));
    sLocalMap.put("es",new Locale("es"));
    sLocalMap.put("ru",new Locale("ru"));
  }




  public static void updateLanguage(Context context) {

    Resources resources = context.getResources();
    Configuration config = resources.getConfiguration();
    config.setLocale(getLocaleByLanguage(context));
    DisplayMetrics dm = resources.getDisplayMetrics();
    resources.updateConfiguration(config, dm);
  }


  public static int getLocalIndex(Context context) {
    int defSelected = 0;
    String defKey = SpHelper.getSharedPreferences(context).getString("pref_app_language", null);
    String[] langKeys = context.getResources().getStringArray(R.array.languages_key);
    if (defKey != null) {
      for (int i = 0; i < langKeys.length; i++) {
        if (TextUtils.equals(defKey, langKeys[i])) {
          defSelected = i;
          break;
        }
      }
    }
    return defSelected;
  }

  public static Locale getLocaleByLanguage(Context context) {
    String language = SpHelper.getSharedPreferences(context).getString("pref_app_language", null);
    if(language == null){
      return sDefaultLocal;
    }
    Locale locale = null;
    return (locale = sLocalMap.get(language)) != null ? locale : sDefaultLocal;
  }


  public static Context attachBaseContext(Context context) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      return updateResources(context);
    } else {
      return context;
    }
  }


  @TargetApi(Build.VERSION_CODES.N)
  private static Context updateResources(Context context) {
    Resources resources = context.getResources();
    Locale locale = getLocaleByLanguage(context);

    Configuration configuration = resources.getConfiguration();
    configuration.setLocale(locale);
    configuration.setLocales(new LocaleList(locale));
    return context.createConfigurationContext(configuration);
  }

}

File: app/src/main/kotlin/com/zzzmode/appopsx/ui/core/LangHelper.kt
package com.zzzmode.appopsx.ui.core

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.os.LocaleList
import android.text.TextUtils
import com.zzzmode.appopsx.R
import java.util.*

/**
 * Created by zl on 2017/6/16.
 */

object LangHelper {

    private val TAG = "LangHelper"

    private val sLocalMap = HashMap<String, Locale>()
    private var sDefaultLocal: Locale

    init {
        sDefaultLocal = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            LocaleList.getDefault().get(0)
        } else {
            Locale.getDefault()
        }

        sLocalMap["zh-cn"] = Locale.SIMPLIFIED_CHINESE
        sLocalMap["zh-tw"] = Locale.TRADITIONAL_CHINESE
        sLocalMap["en"] = Locale.ENGLISH
        sLocalMap["cs"] = Locale("cs", "CZ")
        sLocalMap["es"] = Locale("es")
        sLocalMap["ru"] = Locale("ru")
    }


    fun updateLanguage(context: Context) {

        val resources = context.resources
        val config = resources.configuration
        config.setLocale(getLocaleByLanguage(context))
        val dm = resources.displayMetrics
        resources.updateConfiguration(config, dm)
    }


    fun getLocalIndex(context: Context): Int {
        var defSelected = 0
        val defKey = SpHelper
                .getSharedPreferences(context).getString("pref_app_language", null)
        val langKeys = context.resources.getStringArray(R.array.languages_key)
        if (defKey != null) {
            for (i in langKeys.indices) {
                if (TextUtils.equals(defKey, langKeys[i])) {
                    defSelected = i
                    break
                }
            }
        }
        return defSelected
    }

    private fun getLocaleByLanguage(context: Context): Locale? {
        val language = SpHelper
                .getSharedPreferences(context).getString("pref_app_language", null)
                ?: return sDefaultLocal
        return sLocalMap[language] ?: sDefaultLocal
    }


    fun attachBaseContext(context: Context): Context {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            updateResources(context)
        } else {
            context
        }
    }


    @TargetApi(Build.VERSION_CODES.N)
    private fun updateResources(context: Context): Context {
        val resources = context.resources
        val locale = getLocaleByLanguage(context)

        val configuration = resources.configuration
        configuration.setLocale(locale)
        configuration.locales = LocaleList(locale!!)
        return context.createConfigurationContext(configuration)
    }

}
